### ISSUE TITLE
deps: upgrade deps to allow react@^17

### ISSUE THREAD

**Issue by williaster** [airbnb/visx#872]
`react@17` [is out and doesn't contain breaking features](https://reactjs.org/blog/2020/08/10/react-v17-rc.html) so we should update semantic version ranges for `react` `peerDependencies` to allow an upper range of `^17`

**Comment by hshoff** [comment]
No new features, but there are breaking changes to be mindful of:
- https://reactjs.org/blog/2020/08/10/react-v17-rc.html#fixing-potential-issues
- https://reactjs.org/blog/2020/08/10/react-v17-rc.html#other-breaking-changes

Other related issues that could be separate issues or tagged on to this one:
1. Fix [strict mode](https://reactjs.org/docs/strict-mode.html) warnings
1. Drop React 15 support

**Comment by hshoff** [comment]
v17 release announcement: https://reactjs.org/blog/2020/10/20/react-v17.html

**Comment by m0t0r** [comment]
@williaster @hshoff I would like to help with this issue, is this something still available or it's already being worked on ? Also, could you provide some details what exactly we expect to close this issue ? Is it only adding `"react": "^15.0.0-0 || ^16.0.0-0 || ^17.0.0"` or something else ?

**Comment by sarathps93** [comment]
@williaster @hshoff I would like to work on this issue. Shall I pick it up?

**Comment by hshoff** [comment]
Hi @sarathps93, @m0t0r has first dibs. Will leave it up to him if he wants to continue working on this.

-----

> is this something still available or it's already being worked on?

yes.

> Also, could you provide some details what exactly we expect to close this issue ? Is it only adding "react": "^15.0.0-0 || ^16.0.0-0 || ^17.0.0" or something else?

that's a good start, it would be ` || ^17.0.0-0"`. There would be some additional tasks related to this.

Additional tasks:
1. [ ] address v17 breaking changes: https://reactjs.org/blog/2020/08/10/react-v17-rc.html#fixing-potential-issues, https://reactjs.org/blog/2020/08/10/react-v17-rc.html#other-breaking-changes
1. [ ] bump react-dom 
1. [ ] bump associated `@types` deps
1. [ ] drop react 15 support
1. [ ] identify and fix react strict mode warnings
1. [ ] verify tests pass with v17
1. [ ] verify @visx/demo runs without issue on v17

**Comment by m0t0r** [comment]
Thanks @hshoff I will work on this issue. Sorry @sarathps93

**Comment by sarathps93** [comment]
No issues at all @m0t0r :)

**Comment by m0t0r** [comment]
Upgrading to React v17 made 34 tests to fail with same error:

```
    Error name:    "TypeError"
    Error message: "Cannot read property 'child' of undefined"
```
I found out that it's related to enzyme is not yet supporting React 17. There is an [open issue](https://github.com/enzymejs/enzyme/issues/2429). We would have to wait until they release `enzyme-adapter-react-17` which according to [this](https://github.com/enzymejs/enzyme/issues/2429#issuecomment-713251132) comment is still in the works.

**Comment by m0t0r** [comment]
@williaster @hshoff I have a few questions about testing approaches in this project. I understand Airbnb probably has its coding tools standards but I was wondering how strict they are ? :) For example, would it be possible to migrate from enzyme to a much better [Testing Library](https://testing-library.com). I also  noticed use of shallow rendering in tests which is [not the best approach](https://kentcdodds.com/blog/why-i-never-use-shallow-rendering). There is also [cypress-react-unit-test](https://github.com/bahmutov/cypress-react-unit-test) which for projects like `visx` could be a nice option to consider as well, maybe not for all but some of the interactive tests because it renders in real browser. Could you share your thoughts ?

**Comment by yigitlevent** [comment]
Any updates on this? I've been using React 17 for a while now and really want to try visx.

**Comment by liujimj** [comment]
FYI, we've been using `visx` on a React 17 project for a few months now and haven't run into any showstoppers (have simply ignored the peer-dependency warnings).

**Comment by boy51** [comment]
> Any updates on this? I've been using React 17 for a while now in a project and really want to try visx.

I'm running react 17 with visx and installing on npm 7 using `--legacy-peer-deps` which I assume installs react 16.8 as a peer dep in package lock and it works fine.

I too would appreciate if the peer dep is updated though, i don't think much changed in react 17...

**Comment by smallnamespace** [comment]
FYI, I've been using visx + React 17 for the last few months now without issue.

**Comment by raix** [comment]
@m0t0r should not need to update the development dependencies them selfs only open up the peer dependency to "16.x || 17.x"
_(having chained dependencies on enzyme would really lock things down, from the history of enzyme - one of the reasons why developers are moving to RTL - along with better test patterns)_

**Comment by williaster** [comment]
@m0t0r looks like there's an [unofficial `enzyme-adapter-react-17`](https://github.com/enzymejs/enzyme/issues/2429#issuecomment-831006991), not sure if it'd be worth trying that to see if fixes the failing tests you saw.

**Comment by danantal** [comment]
> @m0t0r should not need to update the development dependencies them selfs only open up the peer dependency to "16.x || 17.x"
> _(having chained dependencies on enzyme would really lock things down, from the history of enzyme - one of the reasons why developers are moving to RTL - along with better test patterns)_

I will second this. You don't need to update the development dependencies themselves - but you could still open up for the 17 peer dep :)

**Comment by blueybloke** [comment]
Running with react 17 and next.js 10.2, can confirm it still works if you install with `--legacy-peer-deps`

**Comment by williaster** [comment]
Just an update here that we're going to try to prioritize this in Q3/soon, thanks for all the nudges to keep this a priority.

@gazcn007 and I will try to
- update everything to allow react@17
- (since the [official react-17-adaptor is not released as of this time](https://github.com/enzymejs/enzyme/pull/2430)) update failing tests to `react-testing-library` (so, we will have a mix of tests which we're okay with; over time we'll likely try to migrate to RTL)

We'll also trying to coordinate the breaking changes of the `2.0.0` release with the `react-spring@9` upgrade in #1218 .

@boy51, if you're interested in helping maybe we can circle back with a branch and specific tests that we need to convert to RTL?

**Comment by boy51** [comment]
Sure, I'm not super familiar with enzyme but I'd be glad to contribute

**Comment by jackw** [comment]
@williaster If you or any of the other contributors require any help here please let me know as I'm more than happy to assist.

**Comment by gazcn007** [comment]
hi folks, [this PR](https://github.com/airbnb/visx/pull/1268) is ready for review. I upgraded react to support v16 || v17, rewrote enzyme tests to RTL for ones that use `mount` which is not supported in React v17. 
Please ignore visx-react-spring and visx-xychart, because those two are dependent on the react-spring v8 to v9 upgrade, which is addressed in this PR: https://github.com/airbnb/visx/pull/1277

@williaster @m0t0r @boy51 @jackw @keithyang

**Comment by gazcn007** [comment]
React 17 upgrade PR is green and waiting for a stamp
https://github.com/airbnb/visx/pull/1268

With regard to Harry's comment:
- [x] upgraded all packages to use react v16 || v17
- [x] address v17 breaking changes: https://reactjs.org/blog/2020/08/10/react-v17-rc.html#fixing-potential-issues, https://reactjs.org/blog/2020/08/10/react-v17-rc.html#other-breaking-changes
- [x] bump react-dom
- [x] bump associated @types deps
- [x] drop react 15 support
- [ ] identify and fix react strict mode warnings:
   - not sure if that needs to be fixed in this PR
- [x] verify tests pass with v17
- [x] verify @visx/demo runs without issue on v17

**Comment by williaster** [comment]
closed by #1268
